README.md
# toggle_slider
A Dart / Flutter package to create Toggle widgets with Sliders

[![Build Status](https://img.shields.io/travis/badges/badgerbadgerbadger.svg?style=flat-square)](https://travis-ci.org/badges/badgerbadgerbadger)

## Installation

- Add `toggle_slider` in `pubsec.yaml.`
- Import `package:toggle_slider/toggle_slider.dart` in your project.

## Example

```dart
// code away!

import 'package:flutter/material.dart';
import 'package:toggle_slider/toggle_slider.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Toggle Slider Example',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Toggle Slider Example'),
      ),
      body: Center(
        child: ToggleSlider(
          sliderContainerDecoration: BoxDecoration(
            color: Colors.blue,
            borderRadius: BorderRadius.only(
              bottomLeft: Radius.circular(10.0),
              bottomRight: Radius.circular(10.0),
            ),
          ),
          toggleCardNumber: 4,
          sliderTitle: "Slider",
          divisions: 5,
          minSliderValue: 10,
          maxSliderValue: 50,
          sliderContainerHeight: 80,
          sliderContainerWidth: 400,
          cardColor: Colors.blue,
          cardAspectRatio: 1,
          cardWidth: 90,
          cardPadding: EdgeInsets.only(left: 10),
          cardChild: Column(children: [
            Image.network(
              "https://flutter.dev/assets/flutter-lockup-1caf6476beed76adec3c477586da54de6b552b2f42108ec5bc68dc63bae2df75.png",
              width: 50,
              height: 30,
            ),
            Divider(
              endIndent: 10,
              indent: 10,
              color: Colors.black,
            ),
          ]),
        ),
      ),
    );
  }
}

```

## Screenshots

[![Screenshots](https://i.imgur.com/UzMNWNt.jpg)]()
[![Screenshots](https://i.imgur.com/m0o6yzz.jpg)]()

## HAPPY CODING ... 🔨🔨🔨

CHANGELOG.md
## [0.0.1+1] - 20.09.2020

* Toggle Slider version 0.0.1+1


lib/toggle_card.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

Widget toggleCard({
  @required BuildContext context,
  @required bool isSelected,
  Color cardColor,
  Widget child,
  double aspectRatio,
  double width,
  double height,
  BorderRadiusGeometry borderRadius,
  DecorationImage image,
  EdgeInsetsGeometry padding,
  String sliderValue,
  TextStyle cardValueStyle,
}) {
  return Align(
    alignment: Alignment(1.0, 1.0),
    child: AspectRatio(
      aspectRatio: aspectRatio ?? 1,
      child: Container(
        padding: padding,
        height: height,
        width: width,
        decoration: BoxDecoration(
          image: image,
          borderRadius: borderRadius,
          color: cardColor ?? Colors.grey,
          border: Border(
            bottom: BorderSide(
                style: isSelected ? BorderStyle.none : BorderStyle.solid,
                width: 6,
                color: isSelected ? cardColor ?? Colors.grey : Colors.white),
          ),
        ),
        child: Column(
          children: [
            child ?? SizedBox(),
            Text(
              sliderValue.toString(),
              style: cardValueStyle,
            ),
          ],
        ),
      ),
    ),
  );
}


lib/custom_slider.dart
import 'package:flutter/material.dart';

class TargetTemperatureSlider extends StatelessWidget {
  final double sliderValue;
  final bool visibility;
  final Function() onTap;
  final Function(double) onChanged;
  final String sliderTitle;
  final int divisions;
  final double minSliderValue;
  final double maxSliderValue;
  final double sliderContainerHeight;
  final double sliderContainerWidth;
  final Decoration sliderContainerDecoration;
  final TextStyle sliderTextStyle;

  const TargetTemperatureSlider({
    @required this.sliderValue,
    @required this.visibility,
    @required this.onTap,
    @required this.onChanged,
    @required this.sliderTitle,
    @required this.divisions,
    @required this.minSliderValue,
    @required this.maxSliderValue,
    this.sliderContainerHeight,
    this.sliderContainerWidth,
    this.sliderContainerDecoration,
    this.sliderTextStyle,
  });

  @override
  Widget build(BuildContext context) {
    return Visibility(
      visible: visibility,
      child: Container(
        height: sliderContainerHeight,
        width: sliderContainerWidth,
        decoration: sliderContainerDecoration,
        child: Column(
          children: <Widget>[
            Row(
              children: <Widget>[
                Padding(
                  padding: EdgeInsets.only(left: 20, top: 10),
                  child: Text(
                    sliderTitle,
                    style: Theme.of(context).textTheme.caption.merge(
                          TextStyle(
                            fontSize: MediaQuery.of(context).size.height / 55,
                          ),
                        ),
                  ),
                ),
                Spacer(),
                Padding(
                  padding: EdgeInsets.only(right: 10, top: 10),
                  child: GestureDetector(
                    child: Icon(Icons.close, size: 20),
                    onTap: onTap,
                  ),
                ),
              ],
            ),
            Container(
              child: Row(
                children: <Widget>[
                  Container(
                    width: MediaQuery.of(context).size.width / 1.2,
                    child: Slider(
                      activeColor: Colors.white,
                      inactiveColor: Colors.grey,
                      min: minSliderValue,
                      max: maxSliderValue,
                      divisions: divisions,
                      value: sliderValue ?? minSliderValue,
                      onChanged: onChanged,
                    ),
                  ),
                  Align(
                    alignment: Alignment.topRight,
                    child: Text(
                      sliderValue == null ? "-" : sliderValue.toString(),
                      style: sliderTextStyle,
                    ),
                  )
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/toggle_slider.dart
library toggle_slider;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:toggle_slider/custom_slider.dart';
import 'package:toggle_slider/toggle_card.dart';

class ToggleSlider extends StatefulWidget {
  final int toggleCardNumber;
  final String sliderTitle;
  final int divisions;
  final double minSliderValue;
  final double maxSliderValue;
  final double sliderContainerHeight;
  final double sliderContainerWidth;
  final Decoration sliderContainerDecoration;
  final Widget cardChild;
  final Color cardColor;
  final double cardAspectRatio;
  final double cardWidth;
  final double cardHeight;
  final BorderRadiusGeometry cardBorderRadius;
  final DecorationImage cardImage;
  final EdgeInsetsGeometry cardPadding;
  final TextStyle cardValueStyle;
  final TextStyle sliderTextStyle;

  ToggleSlider({
    Key key,
    this.toggleCardNumber = 1,
    this.sliderTitle,
    this.divisions,
    @required this.minSliderValue,
    @required this.maxSliderValue,
    this.sliderContainerHeight,
    this.sliderContainerWidth,
    this.sliderContainerDecoration,
    this.cardChild,
    this.cardColor,
    this.cardAspectRatio,
    this.cardWidth,
    this.cardBorderRadius,
    this.cardImage,
    this.cardPadding,
    this.cardHeight,
    this.cardValueStyle,
    this.sliderTextStyle,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => ToggleSliderState();
}

class ToggleSliderState extends State<ToggleSlider> {
  List<bool> isSelected;
  bool visibility;

  List<bool> selectedList = [];
  List<double> targetTemperatureList = [];
  int currentPressedIndex;

  @override
  void initState() {
    for (int i = 0; i < widget.toggleCardNumber; i++) {
      selectedList.add(false);
    }
    for (int i = 0; i < widget.toggleCardNumber; i++) {
      targetTemperatureList.add(null);
    }
    visibility = false;
    isSelected = [false, false, false, false];

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        ToggleButtons(
          fillColor: Colors.transparent,
          selectedBorderColor: Colors.black54,
          disabledColor: Colors.black,
          selectedColor: Colors.white,
          color: Colors.black,
          renderBorder: false,
          children: <Widget>[
            for (var i = 0; i < selectedList.length; i++)
              Padding(
                padding: widget.cardPadding,
                child: toggleCard(
                  sliderValue: targetTemperatureList[i] == null
                      ? "-"
                      : targetTemperatureList[i].floorToDouble().toString(),
                  cardValueStyle: widget.cardValueStyle,
                  width: widget.cardWidth,
                  height: widget.cardHeight,
                  cardColor: widget.cardColor,
                  image: widget.cardImage,
                  aspectRatio: widget.cardAspectRatio,
                  context: context,
                  isSelected: selectedList[i],
                  child: widget.cardChild,
                ),
              )
          ],
          onPressed: (int index) {
            setState(() {
              for (var i = 0; i < selectedList.length; i++) {
                if (i != index) {
                  selectedList[i] = false;
                }
              }

              currentPressedIndex = index;
              selectedList[index] = !selectedList[index];
              visibility = selectedList[index];
            });
          },
          isSelected: selectedList,
        ),
        TargetTemperatureSlider(
          sliderContainerDecoration: widget.sliderContainerDecoration,
          sliderContainerHeight: widget.sliderContainerHeight,
          sliderContainerWidth: widget.sliderContainerWidth,
          minSliderValue: widget.minSliderValue,
          maxSliderValue: widget.maxSliderValue,
          divisions: widget.divisions,
          sliderTitle: widget.sliderTitle,
          sliderTextStyle: widget.sliderTextStyle,
          sliderValue: currentPressedIndex != null
              ? targetTemperatureList[currentPressedIndex]
              : null,
          visibility: visibility,
          onTap: () {
            setState(() {
              selectedList[currentPressedIndex] = false;
              visibility = false;
            });
          },
          onChanged: (newTargetTemperature) {
            setState(() {
              targetTemperatureList[currentPressedIndex] = newTargetTemperature;
            });
          },
        ),
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:toggle_slider/toggle_slider.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Toggle Slider Example',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Toggle Slider Example'),
      ),
      body: Center(
        child: ToggleSlider(
          sliderContainerDecoration: BoxDecoration(
            color: Colors.blue,
            borderRadius: BorderRadius.only(
              bottomLeft: Radius.circular(10.0),
              bottomRight: Radius.circular(10.0),
            ),
          ),
          toggleCardNumber: 4,
          sliderTitle: "Slider",
          divisions: 5,
          minSliderValue: 10,
          maxSliderValue: 50,
          sliderContainerHeight: 80,
          sliderContainerWidth: 400,
          cardColor: Colors.blue,
          cardAspectRatio: 1,
          cardWidth: 90,
          cardPadding: EdgeInsets.only(left: 10),
          cardChild: Column(children: [
            Image.network(
              "https://flutter.dev/assets/flutter-lockup-1caf6476beed76adec3c477586da54de6b552b2f42108ec5bc68dc63bae2df75.png",
              width: 50,
              height: 30,
            ),
            Divider(
              endIndent: 10,
              indent: 10,
              color: Colors.black,
            ),
          ]),
        ),
      ),
    );
  }
}
